core/embed/sec/tropic/tropic.c

```diff
@@ -43,6 +43,65 @@
 // Maximum time to wait for Tropic to boot. Chosen arbitrarily.
 #define TROPIC_BOOT_TIMEOUT_MS 1000
 
+// KEK masks used in PIN verification
+#define TROPIC_KEK_MASKS_PRIVILEGED_SLOT 128
+#define TROPIC_KEK_MASKS_UNPRIVILEGED_SLOT 256
+
+// Mac-and-destroy slots used in PIN verification
+#define TROPIC_FIRST_MAC_AND_DESTROY_SLOT_PRIVILEGED 0
+#define TROPIC_FIRST_MAC_AND_DESTROY_SLOT_UNPRIVILEGED 64
+#define TROPIC_MAC_AND_DESTROY_SLOTS_COUNT 64
+
+// The value by which the index of the first mac-and-destroy slot is shifted
+// before every PIN change. The value is coprime to
+// TROPIC_MAC_AND_DESTROY_SLOTS_COUNT to ensure that
+// `get_mac_and_destroy_slot(i, change_pin_counter)` achieves the maximum
+// possible period in `change_pin_counter`.
+#define TROPIC_MAC_AND_DESTROY_SHIFT 11
+
+// The slot that stores the number of times the PIN has been changed.
+// Since the counter can only be decremented, its acutal value is
+// TROPIC_CHANGE_COUNTER_SLOT_MAX_VALUE minus the number of PIN changes. This
+// slot is used both for both privileged and the unprivileged sessions.
+#define TROPIC_CHANGE_COUNTER_SLOT MCOUNTER_INDEX_4
+#define TROPIC_CHANGE_COUNTER_SLOT_MAX_VALUE 0xfffffffe
+
+#ifdef TREZOR_EMULATOR
+#define TROPIC_RETRY_COMMAND(command) command
+#else
+bool tropic_session_start(void);
+
+#define TROPIC_MAX_RETRIES 5
+// Statement expression, see
+// https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html
+#define TROPIC_RETRY_COMMAND(command)                                       \
+  ({                                                                        \
+    lt_ret_t TROPIC_RETRY_COMMAND_res = LT_FAIL;                            \
+    for (int TROPIC_RETRY_COMMAND_i = 0;                                    \
+         TROPIC_RETRY_COMMAND_i < TROPIC_MAX_RETRIES;                       \
+         TROPIC_RETRY_COMMAND_i++) {                                        \
+      TROPIC_RETRY_COMMAND_res = command;                                   \
+      if (TROPIC_RETRY_COMMAND_res == LT_L1_SPI_ERROR) {                    \
+        continue;                                                           \
+      }                                                                     \
+      if (TROPIC_RETRY_COMMAND_res == LT_L1_CHIP_ALARM_MODE) {              \
+        bool session_started = g_tropic_driver.session_started;             \
+        pkey_index_t pairing_key_index = g_tropic_driver.pairing_key_index; \
+        tropic_deinit();                                                    \
+        tropic01_reset();                                                   \
```

Comment:
@TychoVrahe Can you confirm `tropic01_reset()` will power-cycle the tropic chip? We need it to exit alarm mode.

Reply:
it should, do you have any suspicion that it doesn't work properly?